pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -103,48 +115,193 @@ type controllerVolumeTopology struct {
-			if err != nil {
-				log.Errorf("failed to create an informer for CSINodeTopology instances. Error: %+v", err)
-				return nil, err
+				controllerVolumeTopologyInstance = &controllerVolumeTopology{
+					k8sConfig:               config,
+					nodeMgr:                 nodeManager,
+					csiNodeTopologyInformer: *crInformer,
+				}
+				log.Info("Topology service initiated successfully")
 			}
-
-			controllerVolumeTopologyInstance = &controllerVolumeTopology{
-				k8sConfig:               config,
-				nodeMgr:                 nodeManager,
-				csiNodeTopologyInformer: *crInformer,
+		} else {
+			controllerVolumeTopologyInstanceLock.RUnlock()
+		}
+		return controllerVolumeTopologyInstance, nil
+	} else if c.clusterFlavor == cnstypes.CnsClusterFlavorWorkload {
+		controllerVolumeTopologyInstanceLock.RLock()
+		if wcpControllerVolumeTopologyInstance == nil {
+			controllerVolumeTopologyInstanceLock.RUnlock()
+			controllerVolumeTopologyInstanceLock.Lock()
+			defer controllerVolumeTopologyInstanceLock.Unlock()
+			if wcpControllerVolumeTopologyInstance == nil {
+
+				// Get in cluster config for client to API server.
+				config, err := k8s.GetKubeConfig(ctx)
+				if err != nil {
+					log.Errorf("failed to get kubeconfig with error: %v", err)
+					return nil, err
+				}
+				// Create and start an informer on AvailabilityZone instances.
+				azInformer, err := startAvailabilityZoneInformer(ctx, config)
+				if err != nil {
+					log.Errorf("failed to create an informer for CSINodeTopology instances. Error: %+v", err)
+					return nil, err
+				}
+				// If AvailabilityZone CR not registered, no need to initiate the informer.
+				if azInformer == nil {
+					log.Info("Initiating AvailabilityZone informer not required.")
+					return nil, nil
+				}
+				wcpControllerVolumeTopologyInstance = &wcpControllerVolumeTopology{
+					k8sConfig:  config,
+					azInformer: *azInformer,
+				}
 			}
-			log.Info("Topology service initiated successfully")
+		} else {
+			controllerVolumeTopologyInstanceLock.RUnlock()
 		}
-	} else {
-		controllerVolumeTopologyInstanceLock.RUnlock()
+		return wcpControllerVolumeTopologyInstance, nil
+	}
+	return nil, logger.LogNewErrorf(log, "InitTopologyServiceInController not implemented for "+
+		"cluster flavor: %q", c.clusterFlavor)
+}
+
+// startAvailabilityZoneInformer listens on changes to AvailabilityZone instances and updates the azClusterMap cache.
+func startAvailabilityZoneInformer(ctx context.Context, cfg *restclient.Config) (*cache.SharedIndexInformer, error) {
+	log := logger.GetLogger(ctx)
+	// Create an informer for AvailabilityZone instances.
+	dynInformer, err := k8s.GetDynamicInformer(ctx, "topology.tanzu.vmware.com",
+		"v1alpha1", "availabilityzones", metav1.NamespaceAll, cfg, true)
+	if err != nil {
+		// TODO: TKGS-HA - Verify if this works
+		if apiMeta.IsNoMatchError(err) {
+			log.Info("AvailabilityZone CR is not registered on the cluster")
+			return nil, nil
```

Comment:
Return custom error and handle in WCP controller

Reply:
Starting a dynamic informer on a CR which is not registered does not throw an error. Changed this logic to the following:
1. Try to list the AZ CRs first. If we encounter an error, check if it a NoMatchError? If yes, then do not start informer. `c.topologyMgr` will be nil in such cases.
2. While calling List CRs, if we encounter any other error or do not get an error at all, we continue creating a dynamic informer on the CR.

Reply:
Resolved